### PR TITLE
MER RTT Report v2.4  2019 

### PR DESCRIPTION
**Pull request by dbaluku**
https://app.asana.com/0/1133167201254915/1170835482193281/f

### CODE REVIEW

File: api/src/main/java/org/openmrs/module/ugandaemrreports/reports2019/SetupTxRTT2019Report.java [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+import org.openmrs.module.reporting.cohort.definition.CohortDefinition;
+import org.openmrs.module.reporting.common.ObjectUtil;
+import org.openmrs.module.reporting.dataset.definition.CohortIndicatorDataSetDefinition;
+import org.openmrs.module.reporting.evaluation.parameter.Mapped;
+import org.openmrs.module.reporting.evaluation.parameter.Parameter;
+import org.openmrs.module.reporting.indicator.CohortIndicator;
+import org.openmrs.module.reporting.indicator.dimension.CohortDefinitionDimension;
+import org.openmrs.module.reporting.report.ReportDesign;
+import org.openmrs.module.reporting.report.definition.ReportDefinition;
+import org.openmrs.module.ugandaemrreports.library.*;
```
Comment (solemabrothers) [inline api/src/main/java/org/openmrs/module/ugandaemrreports/reports2019/SetupTxRTT2019Report.java:15]: @dbaluku lets remove these wildcard imports

File: api/src/main/java/org/openmrs/module/ugandaemrreports/reports2019/SetupTxRTT2019Report.java [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+            addIndicator(dsd, "9" + key, label, cohortDefinition, "age=between30and34male");
+            addIndicator(dsd, "10" + key, label, cohortDefinition, "age=between35and39male");
+            addIndicator(dsd, "11" + key, label, cohortDefinition, "age=between40and44male");
+            addIndicator(dsd, "12" + key, label, cohortDefinition, "age=between45and49male");
+            addIndicator(dsd, "13" + key, label, cohortDefinition, "age=above50male");
+        }
+    }
+
+
+    public void addIndicator(CohortIndicatorDataSetDefinition dsd, String key, String label, CohortDefinition cohortDefinition, String dimensionOptions) {
```
Comment (solemabrothers) [inline api/src/main/java/org/openmrs/module/ugandaemrreports/reports2019/SetupTxRTT2019Report.java:176]: @dbaluku also is there a possibility to put this method in a publicly shared file such that it is usable accross different reports. we have many reports using this method and we define it in every report.

File: api/src/main/java/org/openmrs/module/ugandaemrreports/reports2019/SetupTxRTT2019Report.java (hunk 1 above)
Comment (dbaluku) [reply api/src/main/java/org/openmrs/module/ugandaemrreports/reports2019/SetupTxRTT2019Report.java]: Done

Comment (dbaluku) [reply api/src/main/java/org/openmrs/module/ugandaemrreports/reports2019/SetupTxRTT2019Report.java]: Done

File: api/src/main/java/org/openmrs/module/ugandaemrreports/reports2019/SetupTxRTT2019Report.java [hunk 3]
```diff
@@ -185,22 +186,33 @@ public void addIndicator(CohortIndicatorDataSetDefinition dsd, String key, Strin
...
                 "(select patient_id,max(encounter_datetime)last_enc_date from encounter where encounter_datetime <=:startDate group by patient_id) last_encounter \n" +
                 "on o.person_id=patient_id where o.concept_id=5096 and o.value_datetime <= :startDate group by person_id)t1 \n " +
-                "where next_visit > last_enc_date and datediff(:startDate,next_visit)>=28 ";
+                "where next_visit > last_enc_date and datediff(:startDate,next_visit)> 28 ";
         SqlCohortDefinition cd = new SqlCohortDefinition(query);
         cd.addParameter(new Parameter("startDate", "startDate", Date.class));
         return  cd;
     }
 
+    public CohortDefinition getLostPatientsDuringPeriodAndReturnToCareWithinSamePeriod() {
```
Comment (solemabrothers) [inline api/src/main/java/org/openmrs/module/ugandaemrreports/reports2019/SetupTxRTT2019Report.java:199]: @dbaluku cant this cohort be generalised for use in other coming reports, looks like its a cohort we can use.

Comment (dbaluku) [reply api/src/main/java/org/openmrs/module/ugandaemrreports/reports2019/SetupTxRTT2019Report.java:199]: @solemabrothers  this cohort is too specific for RTT it cant be used anywhere

File: api/src/main/java/org/openmrs/module/ugandaemrreports/reports2019/SetupTxRTT2019Report.java [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+
+        CohortDefinitionDimension ageDimension = commonDimensionLibrary.getTxNewAgeGenderGroup();
+        dsd.addDimension("age", Mapped.mapStraightThrough(ageDimension));
+
+        CohortDefinition males = cohortDefinitionLibrary.males();
+        CohortDefinition females = cohortDefinitionLibrary.females();
+
+        CohortDefinition onArtDuringQuarter = hivCohortDefinitionLibrary.getPatientsHavingRegimenDuringPeriod();
+        CohortDefinition missedAppointment28DaysAndAbove = df.getPatientsWhoHaveNotComeAfterTheirLastMissedAppointmentByMinimumDays(28);
+        CohortDefinition RTT = df.getPatientsInAll(onArtDuringQuarter,missedAppointment28DaysAndAbove);
```
[the remark this answers is not in the document]
Comment (dbaluku) [reply api/src/main/java/org/openmrs/module/ugandaemrreports/reports2019/SetupTxRTT2019Report.java]: done

File: api/src/main/java/org/openmrs/module/ugandaemrreports/reports2019/SetupTxRTT2019Report.java [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+        CohortIndicator ci = new CohortIndicator();
+        ci.addParameter(ReportingConstants.START_DATE_PARAMETER);
+        ci.addParameter(ReportingConstants.END_DATE_PARAMETER);
+        ci.setType(CohortIndicator.IndicatorType.COUNT);
+        ci.setCohortDefinition(Mapped.mapStraightThrough(cohortDefinition));
+        dsd.addColumn(key, label, Mapped.mapStraightThrough(ci), dimensionOptions);
+    }
+
+
+    public CohortDefinition addParameters(CohortDefinition cohortDefinition) {
```
[the remark this answers is not in the document]
Comment (dbaluku) [reply api/src/main/java/org/openmrs/module/ugandaemrreports/reports2019/SetupTxRTT2019Report.java:210]: this is a one line method

File: api/src/main/java/org/openmrs/module/ugandaemrreports/reports2019/SetupTxRTT2019Report.java (hunk 2 above)
Comment (dbaluku) [reply api/src/main/java/org/openmrs/module/ugandaemrreports/reports2019/SetupTxRTT2019Report.java:180]: Done

**Pull request merged**